assemble.py

```diff
@@ -293,40 +313,71 @@ def get_args():
         found_targets_number = len(targets)
         if found_targets_number == 0:
             logger.warning(err_msg)
+            p.tick(complete=True)
             exit(1)
 
-        p = Progress(len(targets))
         logger.info('Found {} Targets to assemble image for'.format(found_targets_number))
+        apps_root_dir = args.fetch_dir + "/restorable"
+        fetch_progress = Progress(len(targets), p)
+        for target in targets:
+            logger.info(f"Getting info about Target's Lmp release...")
+            release_info = factory_client.get_target_release_info(target)
+            if release_info.lmp_version > 0:
+                logger.info(
+                    f"Target's LmP version: {release_info.lmp_version}, yocto version: {release_info.yocto_version}")
+
+            if args.app_type == 'restorable' or (not args.app_type and release_info.lmp_version > 84):
+                logger.info('Fetching Restorable Apps...')
+                apps_desc = fetch_restorable_apps(target, apps_root_dir, args.app_shortlist, args.token)
+                fetched_apps[target.name] = (apps_desc, os.path.join(args.out_image_dir, target.tags[0]))
+            fetch_progress.tick()
+
+        preload_progress = Progress(2 * len(targets) + len(fetched_apps), p)
         for target in targets:
             logger.info('Assembling image for {}, shortlist: {}'.format(target.name, args.app_shortlist))
-            subprog = Progress(3, p)
             if not target.has_apps():
                 logger.info("Target has no apps, skipping preload")
-                subprog.tick(complete=True)
+                preload_progress.tick(complete=True)
                 continue
-            image_file_path, release_info = factory_client.get_target_system_image(target, args.out_image_dir, subprog)
 
-            apps_dir = None
-            if args.app_type == 'restorable' or (not args.app_type and release_info.lmp_version > 84):
+            try:
+                image_file_path = factory_client.get_target_system_image(target, args.out_image_dir,
+                                                                         preload_progress)
+            except requests.HTTPError as exc:
+                if exc.response.status_code == 404:
+                    # try to download `ota-ext4` image
+                    logger.info("Target's system image in `.wic` format was not found,"
+                                " trying to get an `ext4` image;"
```

Comment:
And probably better to say it is looking for an ota-ext4 image, as a target can also generate a "normal" ext4 image in yocto (without ostree support).

Reply:
Done